File: java/src/main/java/racing/Car.java
package racing;

public class Car {
    private static final int INIT_POSITION = 0;
    private static final int MOVE_THRESHOLD = 4;

    private final CarName name;
    private Position position;

    public Car(String name) {
        this.name = CarName.of(name);
        this.position = Position.of(INIT_POSITION);
    }

    public void move(MovingStrategy movingStrategy) {
        int value = movingStrategy.generateValue();
        if (value >= MOVE_THRESHOLD) {
            position = position.next();
        }
    }

    public CarName getName() {
        return name;
    }

    public Position getPosition() {
        return position;
    }
}

File: ts/src/domain/Car/Car.ts
import { CarName } from '../CarName/CarName';
import { Position } from '../Position/Position';
import { MoveStrategy } from '../strategy/MoveStrategy';

export class Car {
    private readonly name: CarName;
    private position: Position;

    constructor(name: CarName) {
        this.name = name;
        this.position = new Position();
    }

    move(strategy: MoveStrategy): void {
        if (strategy.canMove()) {
            this.position = this.position.advance();
        }
    }

    getName(): CarName {
        return this.name;
    }

    getPosition(): Position {
        return this.position;
    }
}
